src/provider.rs

```diff
@@ -174,6 +197,164 @@ impl TableProvider for SeafowlTable {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SeafowlPruningStatistics {
+    pub partition_count: usize,
+    pub schema: SchemaRef,
+    pub partitions: Arc<Vec<SeafowlPartition>>,
+    pub min_values: HashMap<String, Vec<ScalarValue>>,
+    pub max_values: HashMap<String, Vec<ScalarValue>>,
+    pub null_counts: HashMap<String, Vec<Option<u64>>>,
+}
+
+impl SeafowlPruningStatistics {
+    // Generate maps of min/max/null stats per column, parsing the serialized values stored in partitions
+    fn from_partitions(
+        partitions: Vec<SeafowlPartition>,
+        schema: SchemaRef,
+    ) -> Result<Self> {
+        let partition_count = partitions.len();
+        let mut min_values = HashMap::new();
+        let mut max_values = HashMap::new();
+        let mut null_counts = HashMap::new();
+
+        for field in schema.fields() {
+            min_values.insert(
+                field.name().clone(),
+                vec![ScalarValue::Utf8(None); partition_count],
+            );
+            max_values.insert(
+                field.name().clone(),
+                vec![ScalarValue::Utf8(None); partition_count],
+            );
+            null_counts.insert(field.name().clone(), vec![None; partition_count]);
+        }
+
+        for (ind, partition) in partitions.iter().enumerate() {
+            for column in partition.columns.as_slice() {
+                let data_type = schema
+                    .field_with_name(column.name.deref())
+                    .unwrap()
+                    .data_type();
+
+                min_values.get_mut(column.name.as_ref()).unwrap()[ind] =
+                    Self::parse_bytes_value(&column.min_value, data_type)?;
+                max_values.get_mut(column.name.as_ref()).unwrap()[ind] =
+                    Self::parse_bytes_value(&column.max_value, data_type)?;
+                null_counts.get_mut(column.name.as_ref()).unwrap()[ind] =
+                    column.null_count;
+            }
+        }
+
+        Ok(Self {
+            partition_count,
+            schema,
+            partitions: Arc::new(partitions),
+            min_values,
+            max_values,
+            null_counts,
+        })
+    }
+
+    // Try to deserialize min/max statistics stored as raw bytes
+    fn parse_bytes_value(
+        bytes_value: &Arc<Option<Vec<u8>>>,
+        data_type: &DataType,
+    ) -> Result<ScalarValue> {
+        match bytes_value.as_ref() {
+            Some(bytes) => match String::from_utf8(bytes.clone()) {
+                Ok(string_val) => ScalarValue::try_from_string(string_val, data_type),
+                Err(error) => Err(DataFusionError::Internal(format!(
+                    "Failed to parse min/max value: {}",
+                    error
+                ))),
+            },
+            None => Ok(ScalarValue::Utf8(None)),
```

Comment:
Will this None get treated as a legitimate min/max value by the pruning or will it be ignored?

Reply:
If you're asking how does DF pruning treat None values—it treats them as unknowns, so if it's unsure that a given partition is needed it will include it (e.g. if the min max interval of a partition is `[3, None]`, it will be included for the `s2 > 5` qualifier as some rows may be required, see  [here](https://github.com/apache/arrow-datafusion/blob/3eb55e9a0510d872f6f7765b1a5f17db46486e45/datafusion/core/src/physical_optimizer/pruning.rs#L1591-L1606)).

If you're asking whether `ScalarValue::Utf8(None)` will get converted to `None` in the final min/max array of a given type, I'm unsure but plan to cover it with tests.

Reply:
Indeed, the parsing was wrong; it's [fixed now](https://github.com/splitgraph/seafowl/pull/67/commits/7761b31d6d32e1c3a8b05c4f43489a69140593de), alongside with some (doc)tests that demonstrate it.